src/renderer/stores/sequencer.ts

```diff
@@ -76,23 +141,360 @@ export const useSequencerStore = create<State & Actions>()(
+            currentIdx,
+            state,
+          );
+          state.sequences[currentIdx] = currentSequence;
+        }
+        // Load the previous cycle
+        state.cycleConfig.ptrCycle = state.cycleConfig.ptrCycle + 1;
+        state.sequences = state.cycleRuns[state.cycleConfig.ptrCycle];
+        // Load the first sequence (without saving the previous one)
+        loadSequence(currentIdx, state);
+      }),
+    clearPreviousCycles: () =>
+      set((state) => {
+        state.cycleConfig.ptrCycle = -1;
+        state.cycleRuns = [];
       }),
   })),
 );
+
+// Helper functions ==========================================================
+
+function containerizeCurrentSequence(
+  containerIdx: number,
+  state: State & Actions,
+): TestSequenceContainer {
+  const container = {
+    project: { ...state.testSequenceDisplayed! },
+    tree: { ...state.testSequenceStepTree },
+    elements: [...state.elements],
+    testSequenceUnsaved: state.testSequenceUnsaved,
+    status: state.sequences[containerIdx].status,
+    runable: state.sequences[containerIdx].runable,
+  };
+  return container;
+}
+
+function loadSequence(idx: number, state: State & Actions): void {
+  if (idx < 0 || idx >= state.sequences.length) {
+    return;
+  }
+  state.testSequenceDisplayed = state.sequences[idx].project;
+  state.testSequenceStepTree = state.sequences[idx].tree;
+  state.elements = state.sequences[idx].elements;
+  state.testSequenceUnsaved = state.sequences[idx].testSequenceUnsaved;
+}
+
+function clearSequencer(state: State & Actions): void {
+  state.testSequenceDisplayed = null;
+  state.testSequenceStepTree = {
+    type: "root",
+    children: [],
+    identifiers: [],
+  };
+  state.cycleRuns = [];
+  state.elements = [];
+  state.cycleConfig = {
+    infinite: false,
+    cycleCount: 1,
+    ptrCycle: -1,
+  };
+  state.testSequenceUnsaved = false;
+}
+
+function resetSequencesToPending(state: State & Actions): void {
+  // Clean up all containers
+  state.sequences.forEach((seq: TestSequenceContainer) => {
+    // Clean up the sequence
+    const newElems: TestSequenceElement[] = [...seq.elements].map((elem) => {
+      return elem.type === "test"
+        ? {
+            ...elem,
+            status: "pending",
+            completionTime: undefined,
+            isSavedToCloud: false,
+          }
+        : { ...elem };
+    });
+    seq.elements = newElems;
+    seq.status = "pending";
+    if (seq.project.name === state.testSequenceDisplayed?.name) {
+      state.elements = newElems;
+      state.testSequenceStepTree = seq.tree;
+    }
+  });
+
+  // Clean up the current display
+  const newElems = [...state.elements].map((elem) => {
+    return elem.type === "test"
+      ? {
+          ...elem,
+          status: StatusType.parse("pending"),
+          completionTime: undefined,
+          isSavedToCloud: false,
+        }
+      : { ...elem };
+  });
+  state.elements = newElems;
+}
```

Comment:
since these don't use any hooks, you can just make them store actions which can be called by accessing them through `get()`,  then you don't need to pass in the state explicitly. 
alternatively, if you don't want to expose these functions outside of the store, you can get a setter to the state with `useSequencerStore.setState((state) => ...)`

Reply:
I don't want to expose them, but I the `useSequencerStore.setState((state) => ...)`, right now we don't know which one need a setter and which one is ok with a `get()`
I'll update them